File: clients/scala/src/main/scala/ai_cup_22/debugging/Color.scala
package ai_cup_22.debugging

import ai_cup_22.util.StreamUtil

/**
 * RGBA Color
 *
 * @param r Red component
 * @param g Green component
 * @param b Blue component
 * @param a Alpha (opacity) component
 */
case class Color(r: Double, g: Double, b: Double, a: Double) {
    /**
     * Write Color to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeDouble(stream, r)
        StreamUtil.writeDouble(stream, g)
        StreamUtil.writeDouble(stream, b)
        StreamUtil.writeDouble(stream, a)
    }

    /**
     * Get string representation of Color
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Color { ")
        stringBuilder.append("r: ")
        stringBuilder.append(r)
        stringBuilder.append(", ")
        stringBuilder.append("g: ")
        stringBuilder.append(g)
        stringBuilder.append(", ")
        stringBuilder.append("b: ")
        stringBuilder.append(b)
        stringBuilder.append(", ")
        stringBuilder.append("a: ")
        stringBuilder.append(a)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Color {
    /**
     * Read Color from input stream
     */
    def readFrom(stream: java.io.InputStream): Color = Color(
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream)
    )
}

File: clients/swift/Sources/AiCup22/Debugging/Color.swift
/// RGBA Color
public struct Color {
    /// Red component
    let r: Double

    /// Green component
    let g: Double

    /// Blue component
    let b: Double

    /// Alpha (opacity) component
    let a: Double

    /// Read Color from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Color {
        var r: Double
        r = stream.readDouble()
        var g: Double
        g = stream.readDouble()
        var b: Double
        b = stream.readDouble()
        var a: Double
        a = stream.readDouble()
        return Color(r: r, g: g, b: b, a: a)
    }

    /// Write Color to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeDouble(r)
        stream.writeDouble(g)
        stream.writeDouble(b)
        stream.writeDouble(a)
    }
}
